src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java

```diff
@@ -884,13 +898,28 @@ public StructImpl downloadAndExtract(
             /* ensureNonEmpty= */ !allowFail,
             /* checksumGiven= */ !Strings.isNullOrEmpty(sha256)
                 || !Strings.isNullOrEmpty(integrity));
-    Optional<Checksum> checksum;
-    RepositoryFunctionException checksumValidation = null;
-    try {
-      checksum = validateChecksum(sha256, integrity, urls);
-    } catch (RepositoryFunctionException e) {
-      checksum = Optional.<Checksum>absent();
-      checksumValidation = e;
+    Checksum checksum;
+    
+    if(!sha256.isEmpty() && !integrity.isEmpty()) {
+      /* Terminates the program if both sha256 and integrity are present */
+      throw Starlark.errorf("Expected either 'sha256' or 'integrity', but not both");
+    }
+    else if(!sha256.isEmpty()) {
+      checksum = SuperChecksum.fromString(KeyType.SHA256, sha256).getChecksum();
+    }
+    else if(!integrity.isEmpty()) {
+      try {
+        checksum = Checksum.fromSubresourceIntegrity(integrity);
+      }
+      catch(Checksum.InvalidChecksumException e) {
+        throw new RepositoryFunctionException(Starlark.errorf(
+          "Definition of repository %s: %s at %s",
+          rule.getName(), e.getMessage(), rule.getLocation()),
+          Transience.PERSISTENT);
+      }
+    }
+    else {
+      checksum = null;
```

Comment:
this seems like largely duplicated code that wasn't duplicated in the original. Can you not keep the utility method to avoid duplication?